drivers/i2c/i2c_ite_enhance.c

```diff
@@ -568,6 +605,311 @@ static int i2c_enhance_transfer(const struct device *dev, struct i2c_msg *msgs,
+		irq_enable(config->i2c_irq_base);
+	}
+	/* Wait for the transfer to complete */
+	res = k_sem_take(&data->device_sync_sem, K_MSEC(100));
+
+	irq_disable(config->i2c_irq_base);
+
+	if (res != 0) {
+		data->err = ETIMEDOUT;
+		/* Reset i2c port. */
+		i2c_reset(dev);
+		LOG_ERR("I2C ch%d:0x%X reset cause %d",
+			config->port, data->addr_16bit, I2C_RC_TIMEOUT);
+	}
+
+	i2c_enhance_enable_cpu_idle(config->port);
+
+	return data->err;
+}
+
+static int cq_mode_allowed(const struct device *dev, struct i2c_msg *msgs)
+{
+	struct i2c_enhance_data *data = dev->data;
+
+	/*
+	 * If there is only one message can be got, we can do command queue
+	 * transfer directly.
+	 * If the length of write or read transfer is greater than command
+	 * queue payload size, there will execute PIO mode.
+	 */
+	if (data->num_msgs == 1 && (msgs[0].flags & I2C_MSG_STOP)) {
+		/* e.g. msg.flags = I2C_MSG_READ | I2C_MSG_STOP; */
+		if ((msgs[0].flags & I2C_MSG_READ) &&
+		    (msgs[0].len <= CONFIG_I2C_CQ_MODE_MAX_PAYLOAD_SIZE)) {
+			return 1;
+		/*
+		 * e.g. msg.flags = I2C_MSG_WRITE | I2C_MSG_STOP;
+		 *
+		 * Note: The cmd_i2c_scan function in the i2c_shell.c only
+		 *       writes the address without writing any data (just
+		 *       confirms whether the device exists).
+		 *       The command queue mode is not applicable here, so
+		 *       the PIO mode is used.
+		 */
+		} else if (((msgs[0].flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) &&
+			   (msgs[0].len <= I2C_CQ_MODE_TX_MAX_PAYLOAD_SIZE) &&
+			   (msgs[0].len != 0)) {
+			return 1;
+		} else {
+			return 0;
+		}
+	/*
+	 * If there are two messages, we need to judge whether or not there is
+	 * I2C_MSG_RESTART flag from the second message, and then decide to do
+	 * the command queue or PIO mode transfer.
+	 */
+	} else if (data->num_msgs == 2) {
+		/*
+		 * The first of two messages must be write.
+		 * If the length of write to read transfer is greater than
+		 * command queue payload size, there will execute PIO mode.
+		 */
+		if (((msgs[0].flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) &&
+		    (msgs[0].len <= I2C_CQ_MODE_TX_MAX_PAYLOAD_SIZE)) {
+			/*
+			 * The transfer is i2c_burst_read().
+			 *
+			 * e.g. msg[0].flags = I2C_MSG_WRITE;
+			 *      msg[1].flags = I2C_MSG_RESTART | I2C_MSG_READ |
+			 *                     I2C_MSG_STOP;
+			 */
+			if ((msgs[1].flags & I2C_MSG_RESTART) &&
+			    (msgs[1].flags & I2C_MSG_READ) &&
+			    (msgs[1].flags & I2C_MSG_STOP) &&
+			    (msgs[1].len <= CONFIG_I2C_CQ_MODE_MAX_PAYLOAD_SIZE)) {
+				return 1;
+			/*
+			 * If transfer is i2c_burst_write(), the command queue
+			 * mode is not supported here. We can do the PIO mode
+			 * transfer.
+			 *
+			 * e.g. msg[0].flags = I2C_MSG_WRITE;
+			 *      msg[1].flags = I2C_MSG_WRITE | I2C_MSG_STOP;
+			 */
+			} else {
+				return 0;
+			}
+		} else {
+			return 0;
+		}
+	}
+
+	return 0;
+}
+#endif /* CONFIG_I2C_IT8XXX2_CQ_MODE */
+
+static int i2c_enhance_raw_transfer(const struct device *dev,
```

Comment:
Keeping this as `i2c_enhance_transfer` and calling the PIO function `i2c_enhance_pio_transfer` seems better, since this function isn't a "raw" transfer (it dispatches to a chosen operating mode).

Reply:
Done.